### ISSUE TITLE
PyPy3.9 7.3.10 test regression: tests/test_inspection.py::test_has_attributes[-HasAttributes(a=1, b=2, spam=AnyThing)]

### ISSUE THREAD

**Issue by mgorny** [samuelcolvin/dirty-equals#55]
It seems that PyPy3.9 7.3.10 brings a test regression, compared to 7.3.9:

```pytb
$ /tmp/pypy3.9-v7.3.10-linux64/bin/pypy3 -m pytest
========================================================= test session starts =========================================================
platform linux -- Python 3.9.15[pypy-7.3.10-final], pytest-7.2.0, pluggy-1.0.0
rootdir: /tmp/dirty-equals, configfile: pyproject.toml, testpaths: tests
collected 544 items                                                                                                                   

tests/test_base.py ......s......................                                                                                [  5%]
tests/test_boolean.py ..........................s.................                                                              [ 13%]
tests/test_datetime.py .................................................                                                        [ 22%]
tests/test_dict.py ..........................................................                                                   [ 33%]
tests/test_docs.py ssssssssssssssssssssssssssssssssssssssssssssssssssss                                                         [ 42%]
tests/test_inspection.py ...................F........                                                                           [ 47%]
tests/test_list_tuple.py ..............................................................................                         [ 62%]
tests/test_numeric.py ..........................................................                                                [ 72%]
tests/test_other.py .........................................................................................                   [ 89%]
tests/test_strings.py ...........................................................                                               [100%]

============================================================== FAILURES ===============================================================
____________________________________ test_has_attributes[-HasAttributes(a=1, b=2, spam=AnyThing)] _____________________________________

value = <tests.test_inspection.Foo object at 0x00007f0502d59050>, dirty = HasAttributes(a=1, b=2, spam=AnyThing)

    @pytest.mark.parametrize(
        'value,dirty',
        [
            (Foo(1, 2), HasAttributes(a=1, b=2)),
            (Foo(1, 's'), HasAttributes(a=IsInt(), b=IsStr())),
            (Foo(1, 2), ~HasAttributes(a=IsInt(), b=IsStr())),
            (Foo(1, 2), ~HasAttributes(a=1, b=2, c=3)),
            (Foo(1, 2), HasAttributes(a=1, b=2, spam=AnyThing)),
            (Foo(1, 2), ~HasAttributes(a=1, b=2, missing=AnyThing)),
        ],
        ids=dirty_repr,
    )
    def test_has_attributes(value, dirty):
>       assert value == dirty
E       assert <tests.test_inspection.Foo object at 0x00007f0502d59050> == HasAttributes(a=1, b=2, spam=AnyThing)

tests/test_inspection.py:86: AssertionError
======================================================= short test summary info =======================================================
FAILED tests/test_inspection.py::test_has_attributes[-HasAttributes(a=1, b=2, spam=AnyThing)] - assert <tests.test_inspection.Foo object at 0x00007f0502d59050> == HasAttributes(a=1, b=2, spam=AnyThing)
============================================== 1 failed, 489 passed, 54 skipped in 3.09s ==============================================
```

It seems that `value != expected_value` is `True` for the pair of `<bound method Foo.spam of <tests.test_inspection.Foo object at 0x00007fd043e47600>>` and `AnyThing`. For some reason, `AnyThing.equals()` isn't being used in 7.3.10. I'm afraid I can't debug beyond that because I don't know how it's supposed to work in the first place.

**Comment by cfbolz** [comment]
I've opened a pypy issue to track it on our side: https://foss.heptapod.net/pypy/pypy/-/issues/3879

**Comment by mgorny** [comment]
The issue seems to have been fixed on PyPy end, and I've confirmed that the patch there makes the tests pass for me. I suppose it'll be fixed in 7.3.11 then, unless it turns out to cause even more mayhem.

Feel free to close the bug, though I suppose you're going to hit the problem first hand when CI upgrades to 7.3.10.

**Comment by samuelcolvin** [comment]
same on 3.8, I've skipped it for now.

Closing this issue as tests are passing.
